### PR TITLE
[spirv] Support GatherCmp methods with Status arg.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -2243,41 +2243,70 @@ uint32_t SPIRVEmitter::processTextureGatherRGBACmpRGBA(
...
+    return theBuilder.createImageGather(
+        retType, imageType, image, sampler, coordinate,
+        /*component*/ 0, comparator, constOffset, varOffset, /*constOffsets*/ 0,
+        /*sampleNumber*/ 0, /*isSparse*/ false);
+
+  // If the Status parameter is present, OpImageSparseGather should be used.
+  // The result type of this SPIR-V instruction is a struct in which the first
+  // member is an integer that holds the Residency Code.
+  const auto uintType = theBuilder.getUint32Type();
+  const auto sparseRetType =
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Maybe we should create a method in `TypeTranslator` for this type?

Can handle in the next pull request. I think you'll need it again. ;-P

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -2243,41 +2243,70 @@ uint32_t SPIRVEmitter::processTextureGatherRGBACmpRGBA(
...
+      theBuilder.getStructType({uintType, retType}, "SparseResidencyStruct",
+                               {"Residency.Code", "Result.Type"});
+
+  // Perform ImageSparseGather
+  const auto sparseGather = theBuilder.createImageGather(
+      sparseRetType, imageType, image, sampler, coordinate,
       /*component*/ 0, comparator, constOffset, varOffset, /*constOffsets*/ 0,
-      /*sampleNumber*/ 0, /*isSparse*/ false);
+      /*sampleNumber*/ 0, /*isSparse*/ true);
+  // Write the Residency Code
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Hmm, similar extraction logic like handling GatherRGBA. Extract out to some method?

Also can handle in the next pull request. I think you'll see it again. ;-P

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: You're right. Will do in the next CL.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 2 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: You're right. Will do in the next CL.

**Pull request merged**